### PR TITLE
Python bindings

### PR DESCRIPTION
**Pull request by ryzhyk**
See individual commits.

### CODE REVIEW

File: adapters/src/transport/kafka/input.rs [hunk 1]
```diff
@@ -78,6 +85,45 @@ pub struct KafkaInputConfig {
     group_join_timeout_secs: u32,
 }
 
+// The auto-derived implementation gets confused by the flattened
+// `kafka_options` field. FIXME: I didn't figure out how to attach a
+// `description` to the `topics` property. fields.
```
Comment (mbudiu-vmw) [inline adapters/src/transport/kafka/input.rs:90]: extra dot?

File: pipeline_manager/python/dbsp/error.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
+from dbsp_api_client.types import Response
+from dbsp_api_client.models.error_response import ErrorResponse
+
+class DBSPServerError(Exception):
+    """Exception raised when an HTTP request to the DBSP server fails.
+    """
+
+    def __init__(self, response: Response, description):
```
Comment (mbudiu-vmw) [inline pipeline_manager/python/dbsp/error.py:8]: why not add type annotations for all arguments?

File: pipeline_manager/python/test.py [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+2008-01-02 11:10:21,4521746535378012,208,2,38.05,ee1158a1d18b2bff3ccbcb7f240af2ec,1199301021,41.406169,-88.699498,0
+2008-01-02 11:10:26,3547625380966414,208,2,160.59,7cb077bb10b79c257194775bc7162be0,1199301026,44.57281,-69.605028,0
+2008-01-02 11:10:37,4824771093241,107,2,193.13,4ea3c5e44e2bdf601489bb1fb49f7991,1199301037,40.862917,-73.170228,0
+2008-01-03 05:38:14,4082400842710944,231,2,17.04,abb639d206ea30d766b0fae771cb92e3,1199367494,41.13911,-71.434731,0
+2008-01-04 00:58:43,4082400842710944,208,2,37.78,775c2e16e71852ce2f0f98c14d2aa2cd,1199437123,41.952015,-70.514457,0
+2008-01-04 04:23:04,4082400842710944,381,2,89.9,c0c039baa4d0e62e4b16e0690ac54313,1199449384,42.331448,-70.495846,0
+"""
+
+sql_code = """
+CREATE TABLE demographics (
```
Comment (mbudiu-vmw) [inline pipeline_manager/python/test.py:64]: I think it is nicer to have this in a SQL file. You could just read it from that file.

File: pipeline_manager/src/db.rs [hunk 4]
```diff
@@ -170,50 +169,86 @@ impl ProjectDB {
...
+        unsafe {
+            rusqlite::trace::config_log(Some(|errcode, msg| {
+                log::debug!("sqlite: {msg}:{errcode}")
+            }))?
+        };
+        let dbclient = Connection::open(config.database_file_path())?;
+
+        dbclient.execute(
+            r#"
+CREATE TABLE IF NOT EXISTS project (
```
Comment (mbudiu-vmw) [inline pipeline_manager/src/db.rs:182]: having this code in a function named "connect" sounds wrong

File: pipeline_manager/src/db.rs [hunk 5]
```diff
@@ -222,59 +257,55 @@ impl ProjectDB {
...
-            .await?;
-        let id: ProjectId = ProjectId(row.try_get(0)?);
-
         self.dbclient
             .execute(
-                "INSERT INTO project (id, version, name, code, status_since) VALUES($1, 1, $2, $3, now())",
-                &[&id.0, &project_name, &project_code],
-            )
-            .await?;
+                "INSERT INTO project (version, name, code, status_since) VALUES(1, ?, ?, unixepoch('now'))",
```
Comment (mbudiu-vmw) [inline pipeline_manager/src/db.rs:281]: I didn't know you can use ?. What does it mean?

**Review by mbudiu-vmw — APPROVED** [review]

**Review by lalithsuresh — APPROVED** [review]

File: pipeline_manager/src/db.rs (hunk 5 above)
Comment (lalithsuresh) [reply pipeline_manager/src/db.rs:281]: Parameterized statement that takes values from the tuple in the next line

Comment (ryzhyk) [reply pipeline_manager/src/db.rs:281]: I should use unified syntax for all queries though.  This got messed up in the PG -> SQlite transition

**Pull request merged**